Functionalize and compute joint simultaneously.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88361
* __->__ #88063

This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Functionalize and compute joint simultaneously.

This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
Bah these tests are passing on the branch, need to find the fixes I need

ezyang commented:
it's probably https://github.com/pytorch/pytorch/issues/87575

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
This is blocked on https://github.com/pytorch/pytorch/pull/87647

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -37,22 +37,10 @@ void FunctionalTensorWrapper::set_constructor_metadata() {
-  // we need to know if we're dispatching to AutogradCPU or AutogradXLA).
-  // Instead, it's sufficient to remove the `Dense` dispatch key,
-  // which prevents us from accidentally trying to directly run a CPU/CUDA kernel.
-  key_set_ = key_set_.remove(c10::DispatchKey::Dense);
```
I am curious in what contexts we're querying whether or not a tensor is `Dense` 🤔

Probably user code that asks if the tensor is sparse or not?

ezyang replied:
Sorry, I don't remember, but yeah, in general it was stuff that was looking at the key set to figure out what type of tensor this was. It's too bad we had to get rid of this.

bdhirsh commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -37,22 +37,10 @@ void FunctionalTensorWrapper::set_constructor_metadata() {
   // We override a bunch of _custom(), so make sure they get called
   // TODO: metadata copying may not actually be necessary then
   set_custom_sizes_strides(SizesStridesPolicy::CustomSizes);
+  set_custom_device(true);
```
thanks!

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh commented on c10/core/TensorImpl.h:
```diff
@@ -1307,7 +1307,13 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+    if (is_nested()) {
+      return false;
+    }
+    if (key_set_.has(DispatchKey::Functionalize)) {
```
what ended up causing this to be necessary?

The old behavior seems a bit more correct - if the inner tensor advertises as strided, the outer tensor does too.

ezyang replied:
Oh, this is to get the fix where we don't generate as_strided calls in backwards. I made a policy decision for functionalization to always opt into view meta reconstruction.

ezyang committed:
Update on "Functionalize and compute joint simultaneously."


This also comes with some bug fixes that were uncovered from doing
this:

- Forward device calls to inner tensor in FunctionalTensorWrapper

- Make legacyExtractDispatchKey exclude Functionalize, so that
  it can get at the real device type key.  This is noncontroversial.

- Stop stripping dense from key set.  The reason for this is
  FunctionalWrapperTensor may be used in contexts where people
  query if it is dense or not.  If it doesn't report this correctly
  (from the dispatch key), it will cause errors.  This caused some
  torchbench models to fail when I did one-pass tracing.

- Save and restore reapply views TLS correctly

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -133,7 +132,99 @@ def posthook(grad_input, grad_output):
+            # I do something unsound instead
+            # assert arg is new_arg, "input argument was mutated, this is not valid"
+            if arg is not new_arg:
+                assert arg.shape == new_arg.shape
```
So, there are existing tests in `test_functionalization.py` that mutate input metadata.

To preserve the old behavior, should this PR just do the same thing that `functionalize()` does, and 
 use `.as_strided_()` to mutate the input metadata when detected. That would have to be cleaned up in https://github.com/pytorch/pytorch/pull/82602 when I rebase on top of this PR anyway though.

ezyang replied:
It's a moot point, because dynamo never calls AOTAutograd with ops that mutates inputs, and `as_strided_` is in that set. (And direct user usage is unlikely to call this.)

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1116,7 +1129,6 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
 
     # Deleting this in a followup
     xfail('nn.functional.feature_alpha_dropout', 'with_train'),
-    xfail('nn.functional.pad', 'circular'),
```
Did this PR end up fixing these xfails? 🤔

I thought I fixed this one in particular with https://github.com/pytorch/pytorch/pull/88198. Maybe tracing with functinalization in one go sidesteps this issue.

ezyang replied:
It's fixed by https://github.com/pytorch/pytorch/pull/88063/files#r1014429009

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -416,6 +416,21 @@ def f(x, y):
+        def f(x):
+            return CustomFn.apply(x)
+
+        self.verify_aot_autograd(f, [torch.randn(3)])
```
What's actually going on here, when we pass a custom autograd.function object into `aot_autograd` to be traced?

(1) I thought that "support for custom autograd functions" was pretty unrelated to this change
(2) Does this actually work properly today? (do we trace "through" the custom function, and add its custom fwd and bwd ops into our traced graph)

ezyang replied:
We trace inside the custom autograd function, and end up with an AOTautograd custom function which is identity on backwards and plus one on backwards.

(1) My goal wasn't really to get custom autograd function working (and indeed, in the end to end Dynamo codepath I don't think there's enough wiring for this to work), but one side effect of doing the tracing all in one go is that custom autograd should just work... and it does!
(2) Yes.

bdhirsh commented on functorch/_src/aot_autograd.py:
```diff
@@ -133,7 +132,99 @@ def posthook(grad_input, grad_output):
+    def inner(*args, **kwargs):
+        def to_fun(t):
+            if isinstance(t, Tensor):
+                r = torch._to_functional_tensor(t)
```
I was still thinking about the fact that we're calling `.detach()` now in this PR where previously we didn't have to, and I just want to lay out my understanding.

Basically, if we do `inp_new = inp_old.detach().requires_grad(original_gradness)`, our new input **can** have different than the original input, BUT this metadata only matters in practice w.r.t. mutations. Since we expect to be guaranteed a mutation-free-graph, then these metadata difference won't matter. In particular, the metadata fields that can differ are `._is_leaf` and `._is_view()`.

Are there any tests for aot_dispatch that check that `output._is_leaf` is propagated correctly?

One thing that I'm still not fully clear on is: before this PR, our tensors 

In particular, I'm still a bit unclear on

bdhirsh replied:
For sanity though, are there any tests that exercise aot autograd on inputs that require grad, but are not leaves? Seem worth adding.

ezyang replied:
Yes, the point is that it doesn't matter if you don't mutate inputs.

The non-leaf input tensor case is exercised by some of the end-to-end dynamo tests, though I don't see any direct tests for it with aot autograd. I think my preference is to add the torture tests when we actually support input mutation; that's when those non-leaf/leaf tests are actually interesting.

bdhirsh reviewed (approved):
left a few comments but pre-stamping, generally lgtm

ezyang commented:
@pytorchbot merge
